### DISCUSSION TITLE
Serial TCP bridge bad  split of messages received on software serial

### DISCUSSION THREAD

**Discussion by ouinouin** [arendst/Tasmota#17416] [General]
i m actually using a serial to tcp bridge to connect my Airton Ac to home assistant (see : https://github.com/arendst/Tasmota/discussions/17328 )

i actually managed to decode the frames but the lng frames often comes split : 

```
20:03:34.371 TCP: from MCU: 7A7AD521220000A30A0A1E00000C4B009800000000046401794442313133
20:03:34.383 TCP: from MCU: 373013B2
20:03:34.471 TCP: from MCU: 7A7AD5211C0000A60A0A00000000000000000000000000000000520B
20:03:36.025 WIF: Checking connection...
20:03:44.476 TCP: from MCU: 7A7AD1210D0000A40A0A002525
20:03:54.482 TCP: from MCU: 7A7AD1210D0000A40A0A
20:03:54.492 TCP: from MCU: 002525
```
based on the upper log, the legit messages were : 
7A7AD521220000A30A0A1E00000C4B009800000000046401794442313133373013B2
7A7AD5211C0000A60A0A00000000000000000000000000000000520B
7A7AD1210D0000A40A0A002525
7A7AD1210D0000A40A0A002525

as you can see i got some random splits of the messages , the longuest message i got to receive is : 
7A7AD521220000A30A0A1B00000C4C009800000000046401764442313133373019D6
(34 bytes) , it generally get cut unfortunately.
is there anything to do to not get this split ?  what is triggering this behaviour ? is this the filling of the buffer ? , is there a timeout mechanism ?  in the latest releases i ve seen some things about the serial bridge , but i dont know if it applies ont TCP serial bridge.

**Comment by sfromis** [comment]
IIRC, a steady stream of bytes will be seen as one message, but breaks in the flow is likely to be seen as a split, don't know how short breaks can be without being seen as a new message. For text data, using `SerialDelimiter` can work to control when to see it as break, but that does rarely apply to binary data.

Seeing 10 ms between the last two posted messages is quite a long time for 3 bytes, unless the bit rate is rather low. That looks like uneven transmission of data, reasonable to see as a "break", meaning that the party parsing the data should not rely on that splitting, but instead split based on the lead-in of `7A7A`. This way of doing it is commonly part of how serial protocols are expected to work. In principle, you could use one `7A` as `SerialDelimiter`, but that's "risky" for binary data possibly being able to contain such a single byte.

In case of you not using hardware serial pins, that's recommended for serial running as smoothly as possible.

**Reply by sfromis** [reply]
TLDR: In serial each message is one single byte. Packaging bytes into longer messages takes knowledge of how messages are structured. The very rudimentary chunking based on seeing timing breaks will often be helpful, but does not constitute splitting into identified messages.

**Reply by ouinouin** [reply]
the bitrate is 9600, unfortunately, to my understanding i m using a m5 atom lite, and the software serial port is predated by the usb adapter.
i have to pack the bytes, issue being some messages are longer than others, so i have to find where to gather the results and where not to.
in nodered, while parsing my victron shunt, i was using just a time delay to see different messages, here is seems a bit more complicated.

**Reply by sfromis** [reply]
From what I can see, the messages have a length in the header, which you can use to split, or wait for mere. First of what you decided to see as a message is 34 bytes, and starts with `7A7A D5 21 22 00`, the last of which is 34 when transformed to decimal. The other lines match this pattern too. Maybe the next `00` is part of the length, can't tell for sure. In case of transmission trouble, it is of course possible that any part is broken, but if you check for lead-in of `7A7A` and use the length, you should get close.

**Reply by ouinouin** [reply]
oh man, i completely overlooked that, that was for sure that some bytes represent the length , it will be much easier to code... thanks for that, this is one more step for completeness of this AC unit protocol decoding.

**Comment by barbudor** [comment]
Both serial and TCP are streams that do not handle message split in the transmission layer. Unlike UDP.
So you have to manage the protocol, find start, count etc.. In a continuous stream which may arrive randomly.
Failing to do it (assuming data comes in packets that match the upper layer protocol) is giving the stick to be beaten
